amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -266,11 +275,14 @@ void initialize() {
 
                     TaskResult result = null;
                     if (!skipShardSyncAtWorkerInitializationIfLeasesExist || leaseRefresher.isLeaseTableEmpty()) {
+                        for (int j = 0; j < 10 && leaseRefresher.isLeaseTableEmpty(); j++) {
+                            // check every 1-5 seconds if lease table is still empty,
+                            // to minimize contention between all workers bootstrapping at the same time
+                            long waitTime = ThreadLocalRandom.current().nextLong(1000L, 5000L);
+                            Thread.sleep(waitTime);
+                        }
```

Comment:
This logic will make all the workers to wait for a minimum period of 10 seconds and 50 seconds.  And will make some of the workers to wait for 50 seconds even when the leader completed the sync in say 5 seconds. 
The intention is as follows,

1. Each worker will choose a random time 'T' between 1 and 30 seconds.
2. Each worker will now wait in a loop with 5 second sleep on the exit condition (lease table is still empty and T yet to be reached)
 a. If T is reached on exiting the wait loop, trigger lease sync and proceed to next step. [This means no one else claimed to be leader and am claiming myself to be the leader and updating the lease table]
 b. If T is not reached on exiting the wait loop, proceed to next step [This means some one else claimed to be the leader and am not doing lease sync]
3. Each worker will now wait in a loop with 5 second sleep on the exit condition (hashrange is complete) and with validation (lease table is not empty)
a. The initialization should complete only when the hashrange is complete else it will infinitely block wait in a loop.

Reply:
Also we need unit test cases around this logic.

Reply:
Made the changes. As for unit testing the hashRange check failure, I did not find a neat way to unit test a loop that runs indefinitely. Let me know if you have any ideas.